docs/signing-tool-spec.md

```diff
@@ -0,0 +1,129 @@
+- **Platform:** Authenticode is currently limted to Windows. Thus, while it's possible to sign a NuGet or VSIX cross-platform, the DLL's inside can't be signed unless running on Windows. With the NuGet packages being developer-only artifacts--they're not shipped with the apps--it's critical that the DLL's inside are also signed. Builds for binaries may run on any platform, but as signing is a discreet step in most CI pipelines, it's possible to use a Windows agent for this task.
+
+### Roadmap
+
+The scope of the preview release will be limited to the existing funtionality currently in the service. The remaining functionality in this spec will be delivered for the initial release. The .NET Foundation has a dependency on this tool being delivered by the end of 2022.
+
+#### Preview
+
+**Goals**
+
+- Support for Authenticode, VSIX, NuGet, ClickOnce
+- Only run on Windows x64.
+- Support a single certificate for all files in the operation.
+
+**Non-Goals**
+
+- Strong Name signing won't be in v1; guidance is to use an snk not based on a cert. If easy, perhaps can revisit.
+- Support for Microsoft internal builds and signing services.
+- Containers, including Notary v2 support.
+- Extensibility. v1 will support different signing providers.
+- Support for platforms other than Windows x64. Future work will be required to support ARM64 and non-Windows hosts. Support for certain file types may be limited due to platform support.
+- Offline distribution.
+
+#### v1
+
+**Goals**
+
+- Extensibility mechanism to support different code signing providers with a dynamic lookup so the core client remains agnostic of the backend
+- Offline distribution for core plus backend provider
+- Three providers: Certificate Store, Azure Key Vault, Azure Code Signing
+- Support for additional formats: [.HLKX](https://github.com/dotnet/sign/issues/422), [VBA](https://github.com/dotnet/sign/issues/364)
+- Verification of signatures
+
+
+### Inputs
+
+The parameters to the signing client are as follows:
+
+`sign code AzureKeyVault [options] <files>`
+
+Options:
+
+- `-o | --output` - The output file. If omitted, overwrites input. Must be a directory if multiple files are specified.
+- `-b | --base-directory` - Base directory for files to override the working directory.
+- `-f | --force` - Overwrites a signature if it exists.
+- `-m | --max-concurrency` - Maximum concurrency (default is 4)
+- `-fl | --filelist` - Path to file containing paths of files to sign within an archive
+- `-fd | --file-digest` - The digest algorithm to hash the file with.
+- `-tr | --timestamp-rfc3161` - Specifies the RFC 3161 timestamp server's URL.
+- `-d | --description` - Description
+- `-u | --descriptionUrl` - Description Url
```

Comment:
It seems like details of specific signing types will leak into the option list. For example `description` and `descriptionUrl` seem to be related to ClickOnce and would be unused on a .nupkg. What does the tool do if you provide a description but you're just signing a .nupkg? Will it error out? Ignore the parameter? How is it clear to the person running the command that they shouldn't pass that parameter if they're just signing a .nupkg?

I think more generally, I wonder how this parameter organization scales. It seems like there should be a clear way of denoting parameters that are specific to a certain signing type vs. parameters that apply universally. Example: the `-kv*` parameters seem like general ones but the `--description*` ones seems specific to ClickOnce.

Maybe prefix the parameter with `--click-once-description` and `--click-once-description-url`?

Reply:
The description and description url are used by Authenticode and ClickOnce. I think it's fair to document what common paramters are used by which file types...

That said, they should be used when signing a NuPkg because in that case, they're not just signing the Nupkg but also the DLL's in side, and Authenticode uses those values.

Those common parameters are tied to the code SubCommand whereas the provider-specific ones (all the azure key vault ones) are declared by the AzureKeyVault subcommand. In 1.0, there'd be three providers, `CertificateStore` (in the binary), and externally via the extensibility `AzureKeyVault`, and `AzureCodeSigning`. Each would declare the parameters they need. Additional providers for additional backends would do the same.

Reply:
No all nupkg files have DLLs (e.g. meta packages, content packages). I guess what I'm getting at is that this set of parameters is the union of the signing inputs for ALL artifact types. As the number of artifact types grows, this might become an unwieldy or confusing list. Prefixing the parameter name might be a way to make things a bit more organized or clear how values are being used.

Consider if/when repo signing is added. There are "owners" and "service index" URL metadata values that are embedded in the repository signature. If the parameter is just `--owners` or `--service-index` this may be confusing if you don't plan on really needing NuGet repo signing. However `--nupkg-repo-owners` or `--nupkg-repo-service-index` is a bit clearer.